Add new XML iterator functions

clumens opened the pull request:
This is just a start - there's a ton of places these kinds of loops occur, so I want to get the foundation right before starting in on converting all of them.

kgaillot commented on lib/common/xml.c:
```diff
@@ -3009,6 +3009,21 @@ pcmk__xe_set_props(xmlNodePtr node, ...)
+    CRM_ASSERT(xml != NULL && handler != NULL);
+
+    for (xmlNode *node = pcmk__xe_first_child(xml); node != NULL;
+         node = pcmk__xe_next(node)) {
```
For efficiency's sake I wouldn't use the convenience functions here, it's simple enough to add (node->type == XML_ELEMENT_NODE) in the if

clumens replied:
Fixed.

kgaillot commented on include/crm/common/xml_internal.h:
```diff
@@ -336,4 +336,22 @@ enum xml_private_flags {
+ * \brief Iterate over all child elements of \p xml, performing some
+ *        callback function on each node.
+ *
+ * \param[in] xml                 The starting XML node
```
I would mark it in,out since the handler might change it

kgaillot commented on include/crm/common/xml_internal.h:
```diff
@@ -336,4 +336,22 @@ enum xml_private_flags {
+ *                                child elements will match.
+ * \param[in] handler             The callback function.
+ * \param[in] userdata            User data to pass to the callback function.
+ *                                Can be NULL.
```
ditto

kgaillot commented on lib/common/xml.c:
```diff
@@ -3009,6 +3009,21 @@ pcmk__xe_set_props(xmlNodePtr node, ...)
+                       void (*handler)(xmlNode *xml, void *userdata),
+                       void *userdata)
+{
+    CRM_ASSERT(xml != NULL && handler != NULL);
```
I think xml==NULL should be ok (no-op). That would let us do something like

    pcmk__xe_foreach_child(first_named_child(cib, SOME_ELEMENT), SOME_SUBELEMENT, handler, data));

if SOME_ELEMENT is optional.

kgaillot commented on lib/common/tests/xml/pcmk__xe_foreach_child_test.c:
```diff
@@ -0,0 +1,148 @@
+static void any_of_handler(xmlNode *xml, void *userdata) {
+    function_called();
+    assert_true(pcmk__str_any_of(crm_element_name(xml), "node1", "node2", "node3", NULL));
+}
```
one more test would be worthwhile to ensure we pull out one of these nodes correctly

maybe even a test for each nodeN so we're checking first, middle, last

clumens replied:
Added.

kgaillot commented on include/crm/common/xml_internal.h:
```diff
@@ -354,4 +354,25 @@ pcmk__xe_foreach_child(xmlNode *xml, const char *child_element_name,
+bool
+pcmk__xe_any_child(xmlNode *xml, const char *child_element_name,
+                   bool (*handler)(xmlNode *xml, void *userdata),
+                   void *userdata);
```
I like this better actually, this could be the implementation of pcmk__xe_foreach_child(). It's easy enough to have a handler always return false.

It might be even more flexible to return a standard code, stopping if it's not pcmk_rc_ok. Then the handler could get back a little more info to the caller.

clumens replied:
I think having the handler return a standard code is probably the best approach.  I considered that, but things like `command_delete` in tools/crm_attribute.c have a pretty weird condition on the return code in the loop.  But thinking about it some more, I don't want to build an API around one weird condition.  So I think we can go with the return code and then that one function can just be different.

I also think eventually we're going to want something like:
```
xmlNode *
pcmk__xe_matching_child(xmlNode *xml, const char *child_element_name,
                        bool (*find)(xmlNode *xml, void *userdata),
                        void *userdata);
```

That acts similarly but returns the first child for which `find` returns true.  That sort of thing comes up from time to time in the loops I've looked at.

kgaillot reviewed (commented):
I like it

clumens commented:
Updated to have pcmk__xe_foreach_child and its callback return an int, and dropped the pcmk__xe_any_child patch since it is no longer necessary.  I'm leaving this in draft/WIP for the moment since I haven't yet had time to give it the thorough testing it needs.  But, I think the basics are looking sound now.

kgaillot reviewed (approved):
looks good

clumens commented:
It looks like everything here is covered by existing CTS tests except for the cli_resource_print_cts_constraints function in crm_resource_print.c.  That function is only ever called if you use `crm_resource --list-cts`, which is a hidden option and also requires a colocation constraint to actually print anything.  But then, we already have -A/--stack for that which uses different printing code and is covered by CTS.

kgaillot commented:
> That function is only ever called if you use `crm_resource --list-cts`

fyi that's used by the lab, so it's covered that way

kgaillot commented on lib/common/tests/xml/pcmk__xe_foreach_child_test.c:
```diff
@@ -0,0 +1,197 @@
+    "  </level1>\n"
+    "  <level1>\n"
+    "  </level1>\n"
+    "</xml>";
```
I just realized it would be a good idea to throw some comment and text nodes in the inputs to make sure they're ignored

clumens replied:
Good idea.
